packages/protocol/src/action/request/Read.ts

```diff
@@ -0,0 +1,356 @@
+    DataVersionFilter,
+    EventFilter,
+    EventPath,
+    GlobalAttributes,
+    ReadRequest,
+} from "#types";
+import { camelize } from "@matter/general";
+import { MalformedRequestError } from "./MalformedRequestError.js";
+import { Specifier } from "./Specifier.js";
+
+/**
+ * A read request formulated using Matter protocol semantics.
+ */
+export interface Read extends ReadRequest {}
+
+/**
+ * Formulate a read request using Matter numeric IDs.
+ */
+export function Read(options: Read.Options): Read;
+
+/**
+ * Formulate a read request with extended options and name-based IDs.
+ */
+export function Read(options: Read.Options, ...selectors: Read.Selector[]): Read;
+
+/**
+ * Formulate a read request with name-based IDs.
+ */
+export function Read(...selectors: Read.Selector[]): Read;
+
+export function Read(optionsOrSelector: Read.Options | Read.Selector, ...selectors: Read.Selector[]): Read {
+    let options;
+    if ("kind" in optionsOrSelector) {
+        selectors = [optionsOrSelector, ...selectors];
+        options = {};
+    } else {
+        options = optionsOrSelector;
+    }
+    let { attributes: attributeRequests, versionFilters, events: eventRequests, eventFilters } = options;
+
+    const result: Read = {
+        isFabricFiltered: options.fabricFilter ?? true,
+        interactionModelRevision: options.interactionModelRevision ?? FALLBACK_INTERACTIONMODEL_REVISION,
+    };
+
+    for (const selector of selectors) {
+        reifySelector(selector);
+    }
+
+    if (!attributeRequests?.length && !eventRequests?.length) {
+        throw new MalformedRequestError(`Read action designates no attributes or events`);
+    }
+
+    if (attributeRequests) {
+        result.attributeRequests = attributeRequests;
+    }
+
+    if (versionFilters) {
+        result.dataVersionFilters = versionFilters;
+    }
+
+    if (eventRequests) {
+        result.eventRequests = eventRequests;
+    }
+
+    if (eventFilters !== undefined) {
+        result.eventFilters = eventFilters;
+    }
+
+    return result;
+
+    function reifySelector(selector: Read.Selector) {
+        switch (selector.kind) {
+            case "attribute":
+                reifyAttributeSelector(selector);
+                break;
+
+            case "event":
+                reifyEventSelector(selector);
+                break;
+
+            default:
+                throw new MalformedRequestError(`Invalid selector kind "${(selector as Read.Selector).kind}"`);
+        }
+    }
+
+    /**
+     * Update "real" ReadRequest fields from our convenience attribute "selector".
+     */
+    function reifyAttributeSelector(selector: Read.AttributeSelector) {
+        const cluster = Specifier.clusterOf(selector);
+        const { endpoint } = selector;
+
+        // Install data version filter if the endpoint reports it has complete version information
+        if (typeof endpoint === "object" && typeof endpoint.versions === "object" && typeof cluster === "object") {
```

Comment:
can that be null? (endpoint or cluster)? pot is the isObject tooling method

Reply:
Well, they can't be null according to the types.  Might be good to validate JS clients too... I tweaked so it won't fail here on null but it might elsewhere.

I think we need to revisit `isObject()` because it fails to detect any object with a `toString()` method, so would likely fail here if endpoint is a @matter/node Endpoint.

Reply:
Hmmmm.... need to check that but ok.